src/dmd/lib/elf.d

```diff
@@ -38,10 +38,10 @@ import dmd.root.rmem;
 import dmd.root.string;
 import dmd.root.stringtable;
```

Comment:
From memory, this is one of the big reasons why this isn't a package.

Reply:
I'm not sure I follow your comment here?

Reply:
dmd.root is a logical package, and doesn't depend on anything outside of it save for druntime bindings (i.e another project) - the same goes for dmd.backend as well (this is why there are two OutBuffer types, one in root and the other in backend).

By contrast dmd.lib as a package would import both from its parent dmd package and sibling dmd.root package (so, no encapsulation at all).

Reply:
> By contrast dmd.lib as a package would import both from its parent dmd package and sibling dmd.root package 

So what? 

> (so, no encapsulation at all).

What? That makes no sense. 

Packaging is about:
* grouping modules logically (lib, visitors, AST etc.)
* symbols internal to the package not being exposed (`package` level visibility)
* package level dependence (and more importantly lack thereof, e.g. AST should not depend on semantic analysis).

Reply:
@ibuclaw : It sounds like you are advocating for no cross-package import ? If so, that's pretty unworkable (and IMO not how package were intended to be used).

Reply:
> > (so, no encapsulation at all).
> 
> What? That makes no sense.
> 
> Packaging is about:
> 

This has been repeated ad-nauseam, so might as well just [cross reference](https://forum.dlang.org/post/s8ftnv$aoq$1@digitalmars.com) at this point.

Reply:
> @ibuclaw : It sounds like you are advocating for no cross-package import ? If so, that's pretty unworkable (and IMO not how package were intended to be used).

It is workable, and [it can be done](https://github.com/dlang/dmd/pull/12537) without imports.

Reply:
To be clear, I'm totally for making packages, but I also appreciate the rationale against doing it like this PR.  Agree with it or not, Walter does have a good point here, and you are just going to have to adapt and come up with something that overcomes that disagreement, rather than keep being dismissive about it.

Reply:
For reference, the quote from @WalterBright is:
> To be a little clearer, if the files are all merely reshuffled into various packages, then they all violate the rule:
>
>    *** Never import a file from an uplevel directory ***
>
> and understanding is not increased at all. **And it isn't even just an uplevel directory, it's up then sideways then down.**

Emphasis mine.

This is the silliest rule ever. And it doesn't make sense. Absolutely no code in D (that uses packages) respects it. Look at Phobos, druntime, whatever project, they import sideways. Or is it okay if the top level package is not shared ? This rule is a self-fulfilling restriction for flat package hierarchy, because importing *down* is *completely backwards*.

Packages are trees. The root of the tree is more generic than the leaves, e.g. `std` is more generic than `std.algorithm` which is more generic than `std.algorithm.sorting`.  The abstraction should be at a higher package level, the implementation at a lower level, hence this rule can be essentially rephrased as "the implementation should not import the abstraction".

Reply:
@ibuclaw : Remember https://github.com/sociomantic-tsunami/ocean/issues/573 ?
@donc points to the same issue in the first bullet. Having the abstraction at a lower level than the implementation is nonsense.

Reply:
It's also worth mentioning that dmd/root "was meant as a portable library used by many projects, but it wasn't very good and the only project left using it is dmd". It makes no sense that the lib package is allowed to import `core.stdc.stdlib: strtold` because it's a 'separate project', but not `dmd.root.strtold` because it happens to be in the dmd namespace and violates a "no sideways imports" rule.

Reply:
> It is workable, and [it can be done](https://github.com/dlang/dmd/pull/12537) without imports.

Templatizing something only to import declarations through an intermediate type is still an import, it's just wrapped nicely in gift paper. The dependencies are still there, just hidden by an additional layer.

Reply:
Resolving this